modules/ding_innskraning/ding_innskraning.module

```diff
@@ -0,0 +1,780 @@
+ */
+function _ding_innskraning_auth_id() {
+  // Implementation created from
+  // https://www.texelate.co.uk/blog/create-a-guid-with-php
+  // Create a token.
+  $token = $_SERVER['HTTP_HOST'];
+  $token .= $_SERVER['REQUEST_URI'];
+  $token .= uniqid(rand(), TRUE);
+
+  // GUID is 128-bit hex.
+  $hash = strtoupper(md5($token));
+
+  // Create formatted GUID.
+  $guid = '';
+
+  // GUID format is XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX for readability.
+  $guid .= substr($hash, 0, 8) . '-' . substr($hash, 8, 4) . '-' . substr($hash, 12, 4) .
+    '-' . substr($hash, 16, 4) . '-' . substr($hash, 20, 12);
+
+  return $guid;
+}
+
+/**
+ * Create the login request with SAML parameters.
+ */
+function _ding_innskraning_login_request($id, $issuer, $destination) {
+  // Add the destination URL for login.
+  $destination = url($destination, array('query' => array('id' => $issuer, 'auth' => $id)));
+  $destination = htmlspecialchars($destination);
+  // Take the name of the site as provider name.
+  $provider = variable_get('site_name', '');
+  // Log the time when request was created.
+  $issueInstant = format_date(time(), 'custom', "Y-m-dTH:i:sZ");
+  $providerNameStr = '';
+  // Consumer URL.
+  $consume_url = url('saml/consume', array('absolute' => TRUE));
+
+  if ($provider != '') {
+    $providerNameStr = <<<PROVIDERNAME
+ProviderName="{$provider}"
+PROVIDERNAME;
+  }
+
+  $request = <<<AUTHNREQUEST
+<samlp:AuthnRequest
+  xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol"
+  xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion"
+  ID="$id"
+  Version="2.0"
+  {$providerNameStr}
+  IssueInstant="$issueInstant"
+  Destination="{$destination}"
+  ProtocolBinding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST"
+  AssertionConsumerServiceURL="{$consume_url}">
+  <saml:Issuer>{$issuer}</saml:Issuer>
+    <samlp:NameIDPolicy
+        Format="urn:oasis:names:tc:SAML:2.0:nameid-format:transient"
+        AllowCreate="true" />
+    <samlp:RequestedAuthnContext Comparison="exact">
+        <saml:AuthnContextClassRef>urn:oasis:names:tc:SAML:2.0:ac:classes:X509</saml:AuthnContextClassRef>
+    </samlp:RequestedAuthnContext>
+</samlp:AuthnRequest>
+AUTHNREQUEST;
+
+  $deflatedRequest = gzdeflate($request);
+  $base64Request = base64_encode($deflatedRequest);
+
+  return $base64Request;
+}
+
+/**
+ * Read the key file to sign the request.
+ */
+function _ding_innskraning_get_key($name = FALSE) {
+  $key = NULL;
+  $file_location = drupal_get_path('module', 'ding_innskraning') . '/certs/traustur_bunadur.pem';
+
+  if ($name) {
+    return $file_location;
+  }
+
+  if ($file_location != '') {
+    $key = file_get_contents($file_location);
+  }
+
+  return $key;
+}
+
+/**
+ * Sign the SAML request.
+ */
+function _ding_innskraning_signature($samlRequest, $relayState, $signAlgorithm = XMLSecurityKey::RSA_SHA1) {
+  $key = _ding_innskraning_get_key();
+  if (!$key) {
+    drupal_set_message(t('No key provided, authentication can not be finished.'), 'info');
```

Comment:
This is not a error message suitable for end users. They have no clue about what a key is. If anything this should be handled in a `hook_requirements()` implementations.

Reply:
Yes, but it's not used (the code calling it is commented).